pkg/client/cluster_config.go

```diff
@@ -26,19 +26,21 @@ import (
 
 // NewClusterConfig wraps an existing config's roundtripper
 // with a custom cluster aware roundtripper.
+//
+// Note: it is the caller responsibility to make a copy of the rest config
 func NewClusterConfig(cfg *rest.Config) *rest.Config {
-	copyCfg := rest.CopyConfig(cfg)
-	copyCfg.Wrap(func(rt http.RoundTripper) http.RoundTripper {
+	cfg.Wrap(func(rt http.RoundTripper) http.RoundTripper {
 		return NewClusterRoundTripper(rt)
 	})
 
-	return copyCfg
+	return cfg
 }
 
 // ConfigWithCluster modifies the config host path to include the
 // cluster endpoint.
+//
+// Note: it is the caller responsibility to make a copy of the rest config
 func ConfigWithCluster(cfg *rest.Config, clusterName logicalcluster.Name) *rest.Config {
-	r := rest.CopyConfig(cfg)
-	r.Host = r.Host + clusterName.Path()
-	return r
+	cfg.Host = cfg.Host + clusterName.Path()
```

Comment:
If we're going to proceed with this, I'd probably rename the function to `SetCluster` and not have it return anything, since it's mutating the input.